applications/CompressiblePotentialFlowApplication/custom_processes/apply_far_field_process.cpp

```diff
@@ -0,0 +1,129 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Iñigo Lopez and Marc Núñez
+//
+
+// Project includes
+#include "apply_far_field_process.h"
+
+namespace Kratos {
+
+// Constructor for ApplyFarFieldProcess Process
+ApplyFarFieldProcess::ApplyFarFieldProcess(ModelPart& rModelPart, const double rReferencePotential)
+    : Process(), mrModelPart(rModelPart), mReferencePotential(rReferencePotential)
+{
+    mFreeStreamVelocity = mrModelPart.GetProcessInfo()[FREE_STREAM_VELOCITY];
+}
+
+void ApplyFarFieldProcess::Execute()
+{
+    FindFarthestUpstreamBoundaryNode();
+    AssignFarFieldBoundaryConditions();
+}
+
+void ApplyFarFieldProcess::FindFarthestUpstreamBoundaryNode()
+{
+    // Declaring omp variables, generating vectors of size = num_threads
+    std::size_t num_threads = omp_get_max_threads();
+    std::vector<double> min_projections(num_threads, std::numeric_limits<double>::epsilon());
+    std::vector<std::size_t> nodes_id_list(num_threads, 0);
+
+    //Find minimum node in the direction of the free stream velocity in openmp.
+    double distance_projection=0.0;
+    #pragma omp parallel for firstprivate (distance_projection)
```

Comment:
Absolutely! I implemented it thinking on the 3D case, but Im fine with removing it if it's not worth!

Reply:
I would definitively keep the OpenMP loop. Makes no sense to save a few (in the best of the cases) seconds in the 2D case at the price of making the 3D case slower. Actually, your main computational effort is located in the resolution of the linear system so this wouldn't make any difference in the total computation time.

Furthermore you should take into account that Kratos is conceived to large engineering problems.

Reply:
I agree with both of you. That's why I would optimize the code both for 2D and 3D directly now that we are doing this. My suggestion would be to do something like:

```cpp
void ApplyFarFieldProcess::FindFarthestUpstreamBoundaryNode()
{
    if(mrModelPart.GetProcessInfo()[DOMAIN_SIZE] == 2){
        double min_projection = std::numeric_limits<double>::max();
        for (auto& r_node : mrModelPart.Nodes()){
            const auto& r_coordinates = r_node.Coordinates();
            const double distance_projection = inner_prod(r_coordinates, mFreeStreamVelocity);
            if (distance_projection < min_projection){
                min_projection = distance_projection;
                mpReferenceNode = &r_node;
            }
        }
    }
    else if(mrModelPart.GetProcessInfo()[DOMAIN_SIZE] == 3){
        // Declaring omp variables, generating vectors of size = num_threads
        std::size_t num_threads = omp_get_max_threads();
        std::vector<double> min_projections(num_threads, std::numeric_limits<double>::max());
        std::vector<std::size_t> nodes_id_list(num_threads, 0);
        // Find minimum node in the direction of the free stream velocity in openmp.
        #pragma omp parallel for
        for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++) {
            auto it_node = mrModelPart.NodesBegin() + i;
            std::size_t thread_id = omp_get_thread_num();
            const auto& r_coordinates = it_node->Coordinates();
            const double distance_projection = inner_prod(r_coordinates, mFreeStreamVelocity);
            if (distance_projection < min_projections[thread_id]){
                min_projections[thread_id] = distance_projection;
                nodes_id_list[thread_id] = it_node->Id();
            }
        }

        // Find minimum across all threads
        std::size_t minimum_node_thread_id = 0;
        for (std::size_t i_thread = 0; i_thread<num_threads; i_thread++){
            if (min_projections[i_thread] < min_projections[minimum_node_thread_id]){
                minimum_node_thread_id = i_thread;
            }
        }

        // Storing final result in member variable
        mpReferenceNode = mrModelPart.pGetNode(nodes_id_list[minimum_node_thread_id]);
    }
    else{
        KRATOS_ERROR << "Dimension needs to be 2 or 3." << std::endl;
    }

}
```

Reply:
And what happens with large 2D cases... Honestly I think that it is not worth it to go in this direction. If the case is so small that the parallelization becomes in an overhead, this means that such overhead is negligible due to the small total simulation time.

Reply:
I agree with @rubenzorrilla 
Usually in Kratos we add omp to potentially large loops, since for large cases (this is what Kratos aims at) we definitely need it
You can still control the execution time by setting the number of threads

Reply:
I am good with anything you decide!